README.md
# Audio Modification Library Ryuki Mod Magisk Module
Audio Modification Library (AML) is a compatibility framework originally created by zackptg5 & ahrion that allows the seamless integration of multiple audio mods for Magisk installs. Enables supported audio mods to share the same needed files sudo as audio_effects. [More details in support thread](https://forum.xda-developers.com/apps/magisk/mod-audio-modification-library-t3745466).<br/>
To uninstall, PLEASE REMOVE VIA MAGISK/KERNELSU APP ONLY

## Changelog
* See [Changelog](changelog.md)

## Source Code
* Module [GitHub](https://github.com/Zackptg5/Audio-Modification-Library)

## Ryuki Mod Version Adventages
* Official Magisk v26+ support
* Magisk Delta support
* KernelSU support
* /odm/etc/ & /my_product/etc/ audio files support with Magisk Delta
* /odm/etc/ & /my_product/etc/ audio files mount bind support without Magisk Delta
* stage_policy.conf file support
* Fix bugs & permissions
* Support module that copies audio files via post-fs-data.sh instead of customize.sh
* Does not copy/modify \*audio\*effects\*spatializer\*.xml that causes conflict in audio flinger
* Prevent /data modifying failure in some devices
* Prevent command failure caused by selinux denial
* Does not need to reinstall if Android SDK API version is changed or switched from official Magisk to Magisk Delta and vice-versa

## Ryuki Mod Version Download Link and Changelog
* Update via Magisk/KernelSU app is still directed to official version, so the Ryuki Mod version can only be updated/downloaded from here: https://www.pling.com/p/1981006/




changelog.md
### v5.1 - 7.15.2023
* Fix for magisk delta

### v5.0 - 7.10.2023
* Big thanks to @reiryuki for this update
* Update to current magisk module template - minimum magisk v20.4
* Context fix
* Add stage_policy.conf support
* Add KSU support
* Add support for other partitions - requires ksu or magisk delta
* Rework/simplify some scripting
* Fix for disabling aml in manager

### v4.2 - 1.6.2023
* Add support for V4A RE

### v4.1 - 1.31.2022
* Add support for magisk 24 module update method

### v4.0 - 9.15.2020
* Reworked boot scripts - no more time limitation
* Misc bug fixes
* Removed prop processing
* Allow use of "aml.sh" in addition to ".aml.sh" scripts
* Added support for another dolby mod
* Updated for Android 11
* Only copy files to aml if needed by other mods

### v3.5 - 4.22.2020
* Fixed bug with proxy effect patching

### v3.4 - 4.9.2020
* Fix perms
* Bug fixes

### v3.3 - 03.26.2020
* Fixes for Magisk v20.4

### v3.2 - 03.17.2020
* Restructured script for better efficiency/faster running time - may fix xposed issues
* Fix prop issue where non-aml supported mods were affected
* Favor module .aml.sh scripts over built-ins

### v3.1 - 03.14.2020
* Enhanced osp_detect - needed for some mods
* Added chcon fix for Q - may fix issues some users have with v4a/jdsp
* Use proper contexts for aml files
* Clean up script a little bit

### v3.0 - 03.13.2020
* Complete overhaul - most everything happens in the boot script now, no need for redunduncies
* Added Guitardedhero's Dolby Atmos Pie/Q
* Added Ainur DTS
* Removed IceWizard

### v2.5 - 12.30.2019
* Update to new magisk module template - compatible with magisk 19+ only

### v2.4 - 5.20.2019
* Fix for DTS mod osp effect

### v2.3 - 5.10.2019
* More fixes for system as root devices

### v2.2 - 5.9.2019
* Fix bootmode installation for system as root devices

### v2.1 - 4.3.2019
* Fixed uninstall bug

### v2.0 - 3.28.2019
* Updated to new magisk template - compatible with magisk 18+ only

### v1.9.4 - 2.28.2019
* Bug fixes with mod detection and whatnot

### v1.9.3 - 2.21.2019
* Small bug fix

### v1.9.2 - 12.19.2018
* Rework usb policy patching so it's only for separate module - not applied to add devices by default now
* Various bug fixes

### v1.9.1 - 12.17.2018
* Fixed old bug with upgrades

### v1.9 - 12.16.2018
* Updated for magisk v18 while maintaining backwards compatibility to 15.3 for those problem devices

### v1.8.5 - 12.6.2018
* Added usb fix for v4a/xhifi (and likely other audio apps as well)

### v1.8.4 - 11.28.2018
* Added support for audio_platform_info xml files

### v1.8.3 - 11.5.2018
* Have AML remove patches for disabled mods/not patch disabled mods

### v1.8.2 - 10.22.2018
* Bug fixes for pre_processing patches
* Added capability for multiple pre/post processing patches (See documentation on support thread for usage)
* Proxy entries must include it's own uuid now - allows for custom UUIDs (See documentation on support thread for usage)
* Added Sony Xperia XZ2

### v1.8.1 - 9.2.2018
* Updated to magisk 17 template while maintaining backwards compatibility for magisk 15.3+

### v1.8 - 7.20.2018
* Fix icewizard always showing error
* Add capability to replace libraries and effects with patch_cfgs function (-r)
* Change pre_processing patch_cfgs option to -q
* Fix bug with proxy effects
* Fix bugs with osp with xml files
* Fix bug with squaresound

### v1.7.2 - 7.2.2018
* Updated ice wizard patches
* Delete osp rather than comment out

### v1.7.1 - 6.27.2018
* Bug fix with osp_detect - should patch all cfgs now

### v1.7 - 6.19.2018
* Added support for mixer_gains, audio_device, and sapa_feature xml files
* Added RUNONCE option for custom AML scripts - allows them to only be run once instead of for each audio cfg file. Use this if your mod doesn't have cfg patches
* Added COUNT variable. Can be used by custom AML scripts to determine how many times script has been run. Use this if your mod has cfg patches and other audio file patches
* Fixed old bug with uninstall file restoration
* Fixed bugs with boot script
* Removed need to specify file with patch_cfgs function - make sure you update your aml.sh script for this
* Added ainur squaresound

### v1.6.2 - 6.15.2018
* Fixed mixed up libraries/effects
* Fixed bug with acp

### v1.6.1 - 6.15.2018
* Bug fixes for xml files

### v1.6 - 6.14.2018
* Redid patching backend - huge thanks to Rezmir99 @xda-developers
* Fully integrated aml.sh functionality - users can now user patch_cfgs function and LIBDIR variable like in AudioModificationLibrary.sh script - see support thread for instructions

### v1.5.7 - 4.26.2018
* Add capability to import any audio mod with a ".aml.sh" file in the root of its magisk directory

### v1.5.6 - 4.12.2018
* vendor file fix for devices with separate vendor partitions
* misc fixes

### v1.5.5 - 4.12.2018
* osp_detect fix

### v1.5.4 - 4.08.2018
* dynamic effect removal fix

### v1.5.3 - 4.07.2018
* V4A Fix

### v1.5.2 - 4.07.2018
* Added materialized v4a
* Use dynamic effect removal

### v1.5.1 - 3.30.2018
* Fix effect removals

### v1.5 - 3.28.2018
* Add soundalive and dha effect removal (needed for some samsung devices)
* Pull ACP patch from ACP mod rather than static patch
* Fine tuned osp patching

### v1.4.8 - 3.22.2018
* Replaced ubdr with redone ACP

### v1.4.7 - 3.1.2018

customize.sh
# space
ui_print " "

# log
if [ "$BOOTMODE" != true ]; then
  FILE=/sdcard/$MODID\_recovery.log
  ui_print "- Log will be saved at $FILE"
  exec 2>$FILE
  ui_print " "
fi

# info
MODVER=`grep_prop version $MODPATH/module.prop`
MODVERCODE=`grep_prop versionCode $MODPATH/module.prop`
ui_print " ID=$MODID"
ui_print " Version=$MODVER"
ui_print " VersionCode=$MODVERCODE"
if [ "$KSU" == true ]; then
  ui_print " KSUVersion=$KSU_VER"
  ui_print " KSUVersionCode=$KSU_VER_CODE"
  ui_print " KSUKernelVersionCode=$KSU_KERNEL_VER_CODE"
else
  ui_print " MagiskVersion=$MAGISK_VER"
  ui_print " MagiskVersionCode=$MAGISK_VER_CODE"
fi
ui_print " "

# note
ui_print "- Module detection and patching happens at boot"
ui_print "  The boot script handles everything"
ui_print "  Disabled module will be ignored"
ui_print " "

# Escape each backslash and space since shell will expand it during echo
sed -i -e 's/\\/\\\\/g' -e 's/\ /\\ /g' $MODPATH/AudioModificationLibrary.sh
# Separate AML into individual files for each audio mod
mkdir $MODPATH/.scripts
while read line; do
  case $line in
    \#*) if [ "$uuid" ]; then
           echo " " >> $MODPATH/.scripts/$uuid.sh
         fi
         uuid=$(echo "$line" | sed "s/#//g");;
    *) echo "$line" >> $MODPATH/.scripts/$uuid.sh;;
  esac
done < $MODPATH/AudioModificationLibrary.sh
rm -f $MODPATH/AudioModificationLibrary.sh
# Generate libs var for faster script running
for i in $MODPATH/.scripts/*; do
  libs="$libs-name \"$(basename $i | sed "s/~.*//g")\" "
done
libs="$(echo $libs | sed "s/\" /\" -o /g")"
sed -i -e "s|<libs>|$libs|g" $MODPATH/service.sh

# Set vars in script
[ -z $SERVICED ] && SERVICED=$NVBASE/service.d
amldir=$NVBASE/aml
for i in amldir; do
  for j in post-fs-data service uninstall; do
    sed -i "s|$i=|$i=$(eval echo \$$i)|g" $MODPATH/$j.sh
  done
done

# Place fallback script in the event idiot user deletes aml module in file explorer
mkdir $SERVICED 2>/dev/null
cp -f $MODPATH/uninstall.sh $SERVICED/aml.sh
chmod 0755 $SERVICED/aml.sh
sed -i -e "3a[ -d \"\$moddir/$MODID\" -a ! -f \"\$moddir/$MODID/disable\" ] && exit 0" -e "s|^moddir=.*|moddir=$NVBASE/modules|g" $SERVICED/aml.sh
echo 'rm -f $0' >> $SERVICED/aml.sh

rm -f $MODPATH/install.zip







post-fs-data.sh
# Some devices fails to modify /data at this time without remounting
mount -o rw,remount /data

# Variables
MODPATH="${0%/*}"
amldir=
API="$(getprop ro.build.version.sdk)"
filenames="*audio*effects*.conf -o -name *audio*effects*.xml\
           -o -name *policy*.conf -o -name *policy*.xml\
           -o -name *mixer*paths*.xml -o -name *mixer*gains*.xml\
           -o -name *audio*device*.xml -o -name *sapa*feature*.xml\
           -o -name *audio*platform*info*.xml -o -name *audio*configs*.xml"

# Functions
set_perm() {
  chown $2:$3 $1 || return 1
  chmod $4 $1 || return 1
  CON=$5
  [ -z $CON ] && CON=u:object_r:system_file:s0
  chcon $CON $1 || return 1
}
set_perm_recursive() {
  find $1 -type d 2>/dev/null | while read dir; do
    set_perm $dir $2 $3 $4 $6
  done
  find $1 -type f -o -type l 2>/dev/null | while read file; do
    set_perm $file $2 $3 $5 $6
  done
}
cp_mv() {
  mkdir -p "$(dirname "$3")"
  cp -af "$2" "$3"
  [ "$1" == "-m" ] && rm -f $2 || true
}
osp_detect() {
  local spaces effects type="$1"
  local files=$(find $MODPATH -type f -name "*audio*effects*.conf" -o -name "*audio*effects*.xml")
  for file in $files; do
    for osp in $type; do
      case $file in
        *.conf) spaces=$(sed -n "/^output_session_processing {/,/^}/ {/^ *$osp {/p}" $file | sed -r "s/( *).*/\1/g")
                effects=$(sed -n "/^output_session_processing {/,/^}/ {/^$spaces\$osp {/,/^$spaces}/p}" $file | grep -E "^$spaces +[A-Za-z]+" | sed -r "s/( *.*) .*/\1/g")
                for effect in ${effects}; do
                  spaces=$(sed -n "/^effects {/,/^}/ {/^ *$effect {/p}" $file | sed -r "s/( *).*/\1/g")
                  [ "$effect" != "atmos" -a "$effect" != "dtsaudio" ] && sed -i "/^effects {/,/^}/ {/^$spaces$effect {/,/^$spaces}/d}" $file
                done
                ;;
        *.xml) effects=$(sed -n "/^ *<postprocess>$/,/^ *<\/postprocess>$/ {/^ *<stream type=\"$osp\">$/,/^ *<\/stream>$/ {/<stream type=\"$osp\">/d; /<\/stream>/d; s/<apply effect=\"//g; s/\"\/>//g; s/ *//g; p}}" $file)
                for effect in ${effects}; do
                  [ "$effect" != "atmos" -a "$effect" != "dtsaudio" ] && sed -i "/^\( *\)<apply effect=\"$effect\"\/>/d" $file
                done
                ;;
      esac
    done
  done
  return 0
}

# Debug
exec 2>$MODPATH/debug-pfsd.log
set -x

# Restore and reset
. $MODPATH/uninstall.sh
moddir="$(dirname $MODPATH)" # Changed by uninstall script
rm -rf $amldir $(find $MODPATH/system $MODPATH/vendor -type f) $MODPATH/errors.txt $MODPATH/system.prop 2>/dev/null
[ -f "$moddir/acdb/post-fs-data.sh" ] && mv -f $moddir/acdb/post-fs-data.sh $moddir/acdb/post-fs-data.sh.bak
mkdir $amldir
# Don't follow symlinks
files="$(find /system /odm /my_product -type f -name $filenames)"
for file in $files; do
  name=$(echo "$file" | sed 's|/system||g')
  cp_mv -c $file $MODPATH/system$name
done
files="$(find /vendor -type f -name $filenames)"
if [ -L $MODPATH/system/vendor ]; then
  mkdir -p $MODPATH/vendor
fi
for file in $files; do
  if [ -L $MODPATH/system/vendor ]; then
    cp_mv -c $file $MODPATH$file
  else
    cp_mv -c $file $MODPATH/system$file
  fi
done
rm -f `find $MODPATH -type f -name *audio*effects*spatializer*.xml`
osp_detect "music"

# Detect/move audio mod files
for mod in $(find $moddir/* -maxdepth 0 -type d ! -name aml -a ! -name 'lost+found'); do
  modname="$(basename $mod)"
  [ -f "$mod/disable" ] && continue
  # Move files
  files="$(find $mod -type f -name $filenames 2>/dev/null)"
  [ "$files" ] && echo "$modname" >> $amldir/modlist || continue
  for file in $files; do
    cp_mv -m $file $amldir/$modname$(echo "$file" | sed "s|$mod||g")
  done
  # Chcon fix for Android O+
  if [ "$API" -ge 26 ]; then
    if [ -L $mod/system/vendor ] && [ -d $mod/vendor ]; then
      chcon -R u:object_r:vendor_file:s0 $mod/vendor/lib*/soundfx 2>/dev/null
    else
      chcon -R u:object_r:vendor_file:s0 $mod/system/vendor/lib*/soundfx 2>/dev/null
    fi
  fi
done

# Set perms and such
if [ "$API" -ge 26 ]; then
  set_perm_recursive $MODPATH/system/odm/etc 0 0 0755 0644 u:object_r:vendor_configs_file:s0
  if [ -L $MODPATH/system/vendor ]; then
    set_perm_recursive $MODPATH/vendor 0 2000 0755 0644 u:object_r:vendor_file:s0
    set_perm_recursive $MODPATH/vendor/etc 0 2000 0755 0644 u:object_r:vendor_configs_file:s0
    set_perm_recursive $MODPATH/vendor/odm/etc 0 2000 0755 0644 u:object_r:vendor_configs_file:s0
  else
    set_perm_recursive $MODPATH/system/vendor 0 2000 0755 0644 u:object_r:vendor_file:s0
    set_perm_recursive $MODPATH/system/vendor/etc 0 2000 0755 0644 u:object_r:vendor_configs_file:s0
    set_perm_recursive $MODPATH/system/vendor/odm/etc 0 2000 0755 0644 u:object_r:vendor_configs_file:s0
  fi
fi
exit 0









service.sh
# Variables
MODPATH="${0%/*}"
moddir="$(dirname $MODPATH)"
amldir=
API="$(getprop ro.build.version.sdk)"
[ "$API" -ge 26 ] && libdir="/vendor" || libdir="/system"

# Functions
cp_mv() {
  mkdir -p "$(dirname "$3")"
  cp -af "$2" "$3"
  [ "$1" == "-m" ] && rm -f $2 || true
}
osp_detect() {
  local spaces effects type="$1"
  local files=$(find $MODPATH -type f -name "*audio*effects*.conf" -o -name "*audio*effects*.xml")
  for file in $files; do
    for osp in $type; do
      case $file in
        *.conf) spaces=$(sed -n "/^output_session_processing {/,/^}/ {/^ *$osp {/p}" $file | sed -r "s/( *).*/\1/g")
                effects=$(sed -n "/^output_session_processing {/,/^}/ {/^$spaces\$osp {/,/^$spaces}/p}" $file | grep -E "^$spaces +[A-Za-z]+" | sed -r "s/( *.*) .*/\1/g")
                for effect in ${effects}; do
                  spaces=$(sed -n "/^effects {/,/^}/ {/^ *$effect {/p}" $file | sed -r "s/( *).*/\1/g")
                  [ "$effect" != "atmos" -a "$effect" != "dtsaudio" ] && sed -i "/^effects {/,/^}/ {/^$spaces$effect {/,/^$spaces}/d}" $file
                done
                ;;
        *.xml) effects=$(sed -n "/^ *<postprocess>$/,/^ *<\/postprocess>$/ {/^ *<stream type=\"$osp\">$/,/^ *<\/stream>$/ {/<stream type=\"$osp\">/d; /<\/stream>/d; s/<apply effect=\"//g; s/\"\/>//g; s/ *//g; p}}" $file)
                for effect in ${effects}; do
                  [ "$effect" != "atmos" -a "$effect" != "dtsaudio" ] && sed -i "/^\( *\)<apply effect=\"$effect\"\/>/d" $file
                done
                ;;
      esac
    done
  done
  return 0
}
patch_cfgs() {
  local first=true files="" file lib=false effect=false outsp=false proxy=false replace=false libname libpath effname uid libname_sw uid_sw libname_hw uid_hw libpathsw libpathhw conf xml
  local opt=`getopt :fleoqpr "$@"`
  eval set -- "$opt"
  while true; do
    case "$1" in
      -f) files="placeholder"; shift;;
      -l) lib=true; first=false; shift;;
      -e) effect=true; first=false; shift;;
      -o) outsp=true; conf=output_session_processing; xml=postprocess; first=false; shift;;
      -q) outsp=true; conf=pre_processing; xml=preprocess; first=false; shift;;
      -p) proxy=true; effect=false; outsp=false; first=false; shift;;
      -r) replace=true; shift;;
      --) shift; break;;
      *) return 1;;
    esac
  done
  [ -z "$files" ] && files="$(find $MODPATH -type f -name "*audio*effects*.conf" -o -name "*audio*effects*.xml")" || { files="$1"; shift; }
  $first && { lib=true; effect=true; }
  if $proxy; then
    effname=$1; uid=${2:?}; shift 2
    libname_sw=$1; uid_sw=${2:?}; shift 2
    $lib && { libpathsw=$1; shift; }
    libname_hw=$1; uid_hw=${2:?}; shift 2
    $lib && { libpathhw=${1:?}; shift; }
  else
    $outsp && { type=${1:?}; shift; }
    { $effect || $outsp; } && { effname=${1:?}; shift; }
    $effect && { uid=${1:?}; shift; }
    { $lib || $effect; } && { libname=${1:?}; shift; }
    $lib && { libpath=${1:?}; shift; }
  fi
  for file in $files; do
    case "$file" in
    *.conf)
      if $proxy; then
        if $replace && [ "$(sed -n "/^effects {/,/^}/ {/^  $effname {/,/^  }/p}" $file)" ]; then
          spaces=$(sed -n "/^effects {/,/^}/ {/^ *$effname {/p}" $file | sed -r "s/( *).*/\1/g")
          sed -i "/^effects {/,/^}/ {/^$spaces$effname {/,/^$spaces}/d}" $file
        fi
        [ ! "$(sed -n "/^effects {/,/^}/ {/^  $effname {/,/^  }/p}" $file)" ] && sed -i "s/^effects {/effects {\n  $effname {\n    library proxy\n    uuid $uid\n\n    libsw {\n      library $libname_sw\n      uuid $uid_sw\n    }\n\n    libhw {\n      library $libname_hw\n      uuid $uid_hw\n    }\n  }/g" $file
        if $lib; then
          patch_cfgs -fl "$file" "proxy" "$libdir/lib/soundfx/libeffectproxy.so"
          if $replace; then
            patch_cfgs -frl "$file" "$libname_sw" "$libpathsw"
            patch_cfgs -frl "$file" "$libname_hw" "$libpathhw"
          else
            patch_cfgs -fl "$file" "$libname_sw" "$libpathsw"
            patch_cfgs -fl "$file" "$libname_hw" "$libpathhw"
          fi
        fi
      else
        if $lib; then
          if $replace && [ "$(sed -n "/^libraries {/,/^}/ {/^ *$libname {/,/}/p}" $file)" ]; then
            spaces=$(sed -n "/^libraries {/,/^}/ {/^ *$libname {/p}" $file | sed -r "s/( *).*/\1/g")
            sed -i "/^libraries {/,/^}/ {/^$spaces$libname {/,/^$spaces}/d}" $file
          fi
          [ ! "$(sed -n "/^libraries {/,/^}/ {/^ *$libname {/,/}/p}" $file)" ] && sed -i "s|^libraries {|libraries {\n  $libname {\n    path $libpath\n  }|g" $file
        fi
        if $effect; then
          if $replace && [ "$(sed -n "/^effects {/,/^}/ {/^ *$effname {/,/}/p}" $file)" ]; then
            spaces=$(sed -n "/^effects {/,/^}/ {/^ *$effname {/p}" $file | sed -r "s/( *).*/\1/g")
            sed -i "/^effects {/,/^}/ {/^$spaces$effname {/,/^$spaces}/d}" $file
          fi
          [ ! "$(sed -n "/^effects {/,/^}/ {/^ *$effname {/,/}/p}" $file)" ] && sed -i "s|^effects {|effects {\n  $effname {\n    library $libname\n    uuid $uid\n  }|g" $file
        fi
        if $outsp && [ "$API" -ge 26 ]; then
          local OIFS=$IFS; local IFS=','
          for i in $type; do
            if [ ! "$(sed -n "/^$conf {/,/^}/p" $file)" ]; then
              echo -e "\n$conf {\n    $i {\n        $effname {\n        }\n    }\n}" >> $file
            elif [ ! "$(sed -n "/^$conf {/,/^}/ {/$i {/,/^    }/p}" $file)" ]; then
              sed -i "/^$conf {/,/^}/ s/$conf {/$conf {\n    $i {\n        $effname {\n        }\n    }/g" $file
            elif [ ! "$(sed -n "/^$conf {/,/^}/ {/$i {/,/^    }/ {/$effname {/,/}/p}}" $file)" ]; then
              sed -i "/^$conf {/,/^}/ {/$i {/,/^    }/ s/$i {/$i {\n        $effname {\n        }/}" $file
            fi
          done
          local IFS=$OIFS
        fi
      fi
      ;;
    *.xml)
      if $proxy; then
        if $replace && [ "$(sed -n "/<effects>/,/<\/effects>/ {/^ *<effectProxy name=\"$effname\".*>/,/^ *<\/effectProxy>/p}" $file)" -o "$(sed -n "/<effects>/,/<\/effects>/ {/^ *<effect name=\"$effname\".*\/>/p}" $file)" ]; then
          sed -i "/<effects>/,/<\/effects>/ {/^ *<effectProxy name=\"$effname\".*>/,/^ *<\/effectProxy>/d}" $file
          sed -i "/<effects>/,/<\/effects>/ {/^ *<effect name=\"$effname\".*\/>/d}" $file
        fi
        [ ! "$(sed -n "/<effects>/,/<\/effects>/ {/^ *<effectProxy name=\"$effname\".*>/,/^ *<\/effectProxy>/p}" $file)" -a ! "$(sed -n "/<effects>/,/<\/effects>/ {/^ *<effect name=\"$effname\".*>/,/^ *\/>/p}" $file)" ] && sed -i -e "/<effects>/ a\        <effectProxy name=\"$effname\" library=\"proxy\" uuid=\"$uid\">\n            <libsw library=\"$libname_sw\" uuid=\"$uid_sw\"\/>\n            <libhw library=\"$libname_hw\" uuid=\"$uid_hw\"\/>\n        <\/effectProxy>" $file
        if $lib; then
          patch_cfgs -fl "$file" "proxy" "$libdir/lib/soundfx/libeffectproxy.so"
          if $replace; then
            patch_cfgs -frl "$file" "$libname_sw" "$libpathsw"
            patch_cfgs -frl "$file" "$libname_hw" "$libpathhw"
          else
            patch_cfgs -fl "$file" "$libname_sw" "$libpathsw"
            patch_cfgs -fl "$file" "$libname_hw" "$libpathhw"
          fi
        fi
      else
        if $lib; then
          if $replace && [ "$(sed -n "/<libraries>/,/<\/libraries>/ {/^ *<library name=\"$libname\" path=\"$(basename $libpath)\"\/>/p}" $file)" ]; then
            sed -i "/<libraries>/,/<\/libraries>/ {/^ *<library name=\"$libname\" path=\"$(basename $libpath)\"\/>/d}" $file
          fi
          [ ! "$(sed -n "/<libraries>/,/<\/libraries>/ {/^ *<library name=\"$libname\" path=\"$(basename $libpath)\"\/>/p}" $file)" ] && sed -i "/<libraries>/ a\        <library name=\"$libname\" path=\"$(basename $libpath)\"\/>" $file
        fi
        if $effect; then
          if $replace && [ "$(sed -n "/<effects>/,/<\/effects>/ {/^ *<effect name=\"$effname\".*\/>/p}" $file)" -o "$(sed -n "/<effects>/,/<\/effects>/ {/^ *<effectProxy name=\"$effname\".*>/,/^ *<\/effectProxy>/p}" $file)" ]; then
            sed -i "/<effects>/,/<\/effects>/ {/^ *<effect name=\"$effname\".*\/>/d}" $file
            sed -i "/<effects>/,/<\/effects>/ {/^ *<effectProxy name=\"$effname\".*>/,/^ *<\/effectProxy>/d}" $file
          fi
          [ ! "$(sed -n "/<effects>/,/<\/effects>/ {/^ *<effect name=\"$effname\".*\/>/p}" $file)" ] && sed -i "/<effects>/ a\        <effect name=\"$effname\" library=\"$(basename $libname)\" uuid=\"$uid\"\/>" $file
        fi
        if $outsp && [ "$API" -ge 26 ]; then
          local OIFS=$IFS; local IFS=','
          for i in $type; do
            if [ ! "$(sed -n "/^ *<$xml>/,/^ *<\/$xml>/p" $file)" ]; then
              sed -i "/<\/audio_effects_conf>/i\    <$xml>\n       <stream type=\"$type\">\n            <apply effect=\"$effname\"\/>\n        <\/stream>\n    <\/$xml>" $file
            elif [ ! "$(sed -n "/^ *<$xml>/,/^ *<\/$xml>/ {/<stream type=\"$type\">/,/<\/stream>/p}" $file)" ]; then
              sed -i "/^ *<$xml>/,/^ *<\/$xml>/ s/    <$xml>/    <$xml>\n        <stream type=\"$type\">\n            <apply effect=\"$effname\"\/>\n        <\/stream>/g" $file
            elif [ ! "$(sed -n "/^ *<$xml>/,/^ *<\/$xml>/ {/<stream type=\"$type\">/,/<\/stream>/ {/^ *<apply effect=\"$effname\"\/>/p}}" $file)" ]; then
              sed -i "/^ *<$xml>/,/^ *<\/$xml>/ {/<stream type=\"$type\">/,/<\/stream>/ s/<stream type=\"$type\">/<stream type=\"$type\">\n            <apply effect=\"$effname\"\/>/}" $file
            fi
          done
          local IFS=$OIFS
        fi
      fi
      ;;
    esac
  done
  return 0
}
legacy_script() {
  local RUNONCE=false COUNT=1 LIBDIR=$libdir/lib/soundfx MOD=$mod
  (. $mod/.aml.sh) || echo "Error in $modname aml.sh script" >> $MODPATH/errors.txt
  for file in $files; do
    local NAME=$(echo "$file" | sed "s|$mod|system|g")
    $RUNONCE || { case $file in
                    *audio*effects*) (. $mod/.aml.sh) || [ "$(grep -x "$modname" $MODPATH/errors.txt)" ] || echo "Error in $modname aml.sh script" >> $MODPATH/errors.txt; COUNT=$(($COUNT + 1));;
                  esac; }
  done
}

(
# Debug
exec 2>$MODPATH/debug.log
set -x

# Detect/install audio mods
for mod in $(find $moddir/* -maxdepth 0 -type d ! -name aml); do
  modname="$(basename $mod)"
  [ -f "$mod/disable" ] && continue
  [ -f "$mod/aml.sh" ] && cp -f $mod/aml.sh $mod/.aml.sh
  # .aml.sh file should take precedence
  if [ -f "$mod/.aml.sh" ]; then
    grep -qx "$modname" $amldir/modlist || echo "$modname" >> $amldir/modlist
    if grep -qE '\$MODPATH/\$NAME|RUNONCE=|COUNT=' $mod/.aml.sh; then
      legacy_script
    else
      (. $mod/.aml.sh) || echo "Error in $modname aml.sh script" >> $MODPATH/errors.txt
    fi
  else
    # Favor vendor libs over system ones, no aml builtins are 64bit only - use 32bit lib dir
    if [ -L $MODPATH/system/vendor ]; then
      libs="$(find $mod/vendor/lib/soundfx $mod/system/lib/soundfx -type f <libs> 2>/dev/null)"
    else
      libs="$(find $mod/system/vendor/lib/soundfx $mod/system/lib/soundfx -type f <libs> 2>/dev/null)"
    fi
    for lib in $libs; do
      for audmod in $MODPATH/.scripts/$(basename $lib)~*; do
        uuid=$(basename $audmod | sed -r "s/.*~(.*).sh/\1/g")
        hexuuid="$(echo $uuid | sed -r -e "s/^(..)(..)(..)(..)-(..)(..)-(..)(..)-/\4\3\2\1\6\5\8\7-/g" -e "s/-(..)(..)-(............)$/\2\1\3/g")"
        xxd -p $lib | tr -d '\n' | grep -q "$hexuuid" || continue
        $(grep -xq "$modname" $amldir/modlist 2>/dev/null) || echo "$modname" >> $amldir/modlist
        libfile="$(echo $lib | sed -e "s|$mod||g" -e 's|/system/vendor|/vendor|g')"
        . $audmod
      done
    done
  fi
done

# Reload patched files - original mounted files are seemingly deleted and replaced by sed
for i in $(find $MODPATH/system $MODPATH/vendor -type f); do
  j="$(echo $i | sed -e "s|$MODPATH||g" -e 's|/system/odm|/odm|g' -e 's|/system/my_product|/my_product|g')"
  umount $j
  mount -o bind $i $j
done
[ "$API" -ge 24 ] && killall audioserver || killall mediaserver
exit 0
)&






